test/gas-analytics/HubPool.RefundExecution.ts

```diff
@@ -0,0 +1,180 @@
+      );
+      console.log(`initiateRelayerRefund-cumulativeGasUsed: ${(await initiateTxn.wait()).cumulativeGasUsed}`);
+
+      // Advance time so the request can be executed and execute the request.
+      await timer.setCurrentTime(Number(await timer.getCurrentTime()) + consts.refundProposalLiveness);
+      const txn = await hubPool
+        .connect(dataWorker)
+        .executeRelayerRefund(leaves[leafIndexToExecute], tree.getHexProof(leaves[leafIndexToExecute]));
+
+      // Balances should have updated as expected for tokens contained in the first leaf.
+      for (let i = 0; i < REFUND_TOKEN_COUNT; i++) {
+        expect(await l1Tokens[leafIndexToExecute][i].balanceOf(hubPool.address)).to.equal(
+          STARTING_LP_AMOUNT.sub(SEND_AMOUNT)
+        );
+        expect(await l1Tokens[leafIndexToExecute][i].balanceOf(mockAdapter.address)).to.equal(SEND_AMOUNT);
+      }
+
+      const receipt = await txn.wait();
+      console.log(`executeRelayerRefund-cumulativeGasUsed: ${receipt.cumulativeGasUsed}`);
+    });
+    it("Executing all leaves", async function () {
+      const initiateTxn = await hubPool.connect(dataWorker).initiateRelayerRefund(
+        destinationChainIds, // bundleEvaluationBlockNumbers used by bots to construct bundles. Length must equal the number of leafs.
+        REFUND_CHAIN_COUNT, // poolRebalanceLeafCount. Execute all leaves
+        tree.getHexRoot(), // poolRebalanceRoot. Generated from the merkle tree constructed before.
+        consts.mockDestinationDistributionRoot, // destinationDistributionRoot. Not relevant for this test.
+        consts.mockSlowRelayFulfillmentRoot // Mock root because this isn't relevant for this test.
+      );
+      console.log(`initiateRelayerRefund-cumulativeGasUsed: ${(await initiateTxn.wait()).cumulativeGasUsed}`);
+
+      // Advance time so the request can be executed and execute the request.
+      await timer.setCurrentTime(Number(await timer.getCurrentTime()) + consts.refundProposalLiveness);
+      const txns = [];
+      for (let i = 0; i < REFUND_CHAIN_COUNT; i++) {
+        txns.push(await hubPool.connect(dataWorker).executeRelayerRefund(leaves[i], tree.getHexProof(leaves[i])));
+      }
+
+      // Balances should have updated as expected for tokens contained in the first leaf.
+      for (let i = 0; i < REFUND_CHAIN_COUNT; i++) {
+        for (let j = 0; j < REFUND_TOKEN_COUNT; j++) {
+          expect(await l1Tokens[i][j].balanceOf(hubPool.address)).to.equal(STARTING_LP_AMOUNT.sub(SEND_AMOUNT));
+          expect(await l1Tokens[i][j].balanceOf(mockAdapter.address)).to.equal(SEND_AMOUNT);
+        }
+      }
+
+      // Now that we've verified that the transaction succeeded, let's compute average gas costs.
+      const receipts = await Promise.all(txns.map((_txn) => _txn.wait()));
+      const cumulativeGasUsed = receipts.map((_receipt) => _receipt.cumulativeGasUsed).reduce((x, y) => x.add(y));
```

Comment:
Wherever we're computing gas, I think we should use `gasUsed`, not `cumulativeGasUsed`. The latter is the gas used in this transaction _and all preceding_ transactions in the same block. Not sure this will be a problem in hardhat since it usually maps each transaction to a block, but since that's not _guaranteed_ we probably just want to use gasUsed to be safe.

Source: https://ethereum.stackexchange.com/a/3354/47801